compiler/include/ForallStmt.h

```diff
@@ -0,0 +1,80 @@
+/*
+ * Copyright 2004-2017 Cray Inc.
+ * Other additional copyright holders may be indicated within.
+ *
+ * The entirety of this work is licensed under the Apache License,
+ * Version 2.0 (the "License"); you may not use this file except
+ * in compliance with the License.
+ *
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#ifndef _FORALL_STMT_H_
+#define _FORALL_STMT_H_
+
+#include "stmt.h"
+
+/************************************ | *************************************
+*                                                                           *
+*                                                                           *
+************************************* | ************************************/
+
+//
+// forall loop statement
+//
+class ForallStmt : public Stmt
+{
+public:
+  // access various parts of the node
+  BlockStmt*  masterBlock()    const  { return fMBlock; }
+  bool        zippered()       const  { return fZippered; }
+
+  AList&   iterExprs()                { return fIter; }
+  int      numIterExprs()      const  { return fIter.length; }
+  Expr*    getIterExpr(int i)  const  { return fIter.get(i); }
+  Expr*    firstIterExpr()     const  { return fIter.head; }
+
+  ForallIntents* withClause()    const  { return fWith; }
+  VarSymbol*     parIdxVar()     const; // "chpl__parIdx" variable
+  VarSymbol*     parIdxCopyVar() const; // "chpl__parIdxCopy" variable
+  BlockStmt*     loopBody()      const; // the body of the forall loop
+
+  ForallStmt(BlockStmt* var, Expr* iter, bool zippered,
+             ForallIntents* intents, BlockStmt* body);
+  ~ForallStmt();
+
+  DECLARE_COPY(ForallStmt);
+
+  virtual void        replaceChild(Expr* oldAst, Expr* newAst);
+  virtual void        verify();
+  virtual void        accept(AstVisitor* visitor);
+  virtual GenRet      codegen();
+
+  virtual Expr*       getFirstChild();
+  virtual Expr*       getFirstExpr();
+  virtual Expr*       getNextExpr(Expr* expr);
+
+private:
+  // All pointer fields are always non-null.
+  BlockStmt*     fMBlock;   // Master Block: contains loopBody and defs
+                            // for index variable(s)
+  bool           fZippered;
+  AList          fIter;   // iterable expression(s)
+  ForallIntents* fWith;   // todo: replace with AList of per-variable "clauses"
+};
+
+// serving ForallStmt and forall intents
+bool isDirectlyUnderBlockStmt(const Expr* expr);
+bool isForallStmtIterator(ForallStmt* fs, Expr* expr);
+ForallStmt* enclosingForallStmt(Expr* expr);
+int reduceIntentIdx(ForallStmt* fs, Symbol* reVar);
```

Comment:
If 'var' is listed in the fs's with-clause with a reduce intent, this function returns its position in the with-clause, otherwise -1. 52f91f0 adds a corresonding comment.